### PR TITLE
[shardformer] Align bert value

### LINKED ISSUE #3906: [shardformer] align BERT value
A sharded BERT has been implemented in our shardformer branch. We need to make sure its numerical values can be aligned with the non-sharded version, including the output and gradients.

### PR DESCRIPTION
**Pull request by FoolPlayer**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge

Fixed #3906 

## 📝 What does this PR do?

Add align test for bert, fix a bug in dist loss when meet ignored labels

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/shardformer/test/module_test.py [hunk 1]
```diff
@@ -17,8 +17,10 @@ def get_args():
 
 def test_dist_crossentropy():
     pred = torch.randn(2, 4, 8, requires_grad=True)
-    labels = torch.randint(8, (1, 4)).repeat(2, 1)
-
+    print(pred)
```
Comment (FrankLeeeee) [inline colossalai/shardformer/test/module_test.py:20]: remove print.

File: colossalai/shardformer/test/align_bert.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
+import os
```
Comment (FrankLeeeee) [inline colossalai/shardformer/test/align_bert.py:1]: move this test to `tests/test_shardformer/test_bert.py`.

File: colossalai/shardformer/policies/bert.py [hunk 3]
```diff
@@ -142,7 +142,7 @@ def unembedding() -> List:
                 weight="decoder.weight",
                 bias="decoder.bias",
                 replace_layer=col_nn.Linear1D_Col,
-        # gather_output=True,
+                gather_output=True,
```
Comment (FrankLeeeee) [inline colossalai/shardformer/policies/bert.py:145]: make this configurable from the user api.

File: colossalai/shardformer/layer/dist_crossentropy.py [hunk 4]
```diff
@@ -75,8 +75,8 @@ def forward(ctx, vocab_logits: torch.Tensor, target: torch.Tensor):
 
         # calculate the loss
         # loss = log(sum(exp(x[i]))) - x[class]
-        loss = torch.log(sum_exp_logits) - pred_logits
-        loss = torch.sum(loss).div_(loss.numel())
+        loss = torch.where(target == -100, 0.0, torch.log(sum_exp_logits) - pred_logits)
+        loss = torch.sum(loss).div_(torch.sum(loss != 0.0))
```
Comment (FrankLeeeee) [inline colossalai/shardformer/layer/dist_crossentropy.py:79]: Align this with PyTorch parameter where ignore index is configurable.

File: colossalai/shardformer/policies/bert.py (hunk 3 above)
Comment (FoolPlayer) [reply colossalai/shardformer/policies/bert.py:144]: https://github.com/hpcaitech/ColossalAI/blob/5153e181d749adec09d7455ae26ce3a962607c90/colossalai/shardformer/shard/sharder.py#L153

https://github.com/hpcaitech/ColossalAI/blob/5153e181d749adec09d7455ae26ce3a962607c90/colossalai/shardformer/shard/shard_config.py#L18

gather_output in policy means this layer can be gatherd, gather_output in shardconfig mean whether to gatherd

File: colossalai/shardformer/test/align_bert.py (hunk 2 above)
Comment (FoolPlayer) [reply colossalai/shardformer/test/align_bert.py:1]: done

File: colossalai/shardformer/policies/bert.py [hunk 5]
```diff
@@ -155,7 +155,8 @@ class BertForMaskedLMPolicy(BertPolicy):
 
     @staticmethod
     def inject_policy() -> Tuple[nn.Module, nn.Module]:
-        return (BertForMaskedLM, BertForMaskedLM_)
+        # return (BertForMaskedLM, BertForMaskedLM_)
+        return None
```
Comment (FrankLeeeee) [inline colossalai/shardformer/policies/bert.py:159]: Why return None? The type hinting suggests two items are returned, but only one `None` is returned.

File: colossalai/shardformer/README.md [hunk 6]
```diff
@@ -20,15 +20,20 @@
 The sample API usage is given below:
 
 ``` python
-from colossalai.shardformer import shard_model
+from colossalai.shardformer.shard import ShardConfig, shard_model
```
Comment (FrankLeeeee) [inline colossalai/shardformer/README.md:23]: We should make `ShardConfig` and `shard_model` the top-level API for the user in the `shardformer` module.

e.g. 

```python
from colossalai.shardformer import ShardConfig, shard_model
```
